crates/wasmparser/src/validator.rs

```diff
@@ -492,28 +492,32 @@ impl Validator {
             }
         }
 
-        self.state = match (encoding, num) {
-            (Encoding::Module, WASM_MODULE_VERSION) => {
-                assert!(self.module.is_none());
-                self.module = Some(ModuleState::default());
-                State::Module
+        self.state = match encoding {
+            Encoding::Module => {
+                if num == WASM_MODULE_VERSION {
+                    assert!(self.module.is_none());
+                    self.module = Some(ModuleState::default());
+                    State::Module
+                } else {
+                    bail!(range.start, "unknown binary version: {num:#x}");
+                }
             }
-            (Encoding::Component, WASM_COMPONENT_VERSION) => {
+            Encoding::Component => {
                 if !self.features.component_model {
-                    return Err(BinaryReaderError::new(
-                        "WebAssembly component model feature not enabled",
+                    bail!(
                         range.start,
-                    ));
+                        "unknown binary version: {num:#x}, \
```

Comment:
To match the error messages below:

```suggestion
                        "unknown component version: {num:#x}, \
```

Reply:
Heh I wanted to do that but it runs afoul of [this spec test](https://github.com/WebAssembly/testsuite/blob/4f77306bb63151631d84f58dedf67958eb9911b9/binary.wast#L44) which currently somewhat forces our hand unless specific exemptions for specific errors are added. My laziness won out in this case to acquiesce to the test suite's expectation.